gpu/test/preconditioner/block_jacobi_kernels.cpp

```diff
@@ -107,6 +108,185 @@ class BlockJacobi : public ::testing::Test {
+                {3, 2, 1.0},
+                {3, 3, 1.0},
+                {4, 4, 1.0}}});
+
+    auto d_mtx = Mtx::create(gpu);
+    d_mtx->copy_from(mtx.get());
+
+    std::unique_ptr<BjFactory> bj_factory;
+    bj_factory = BjFactory::create(ref, 3);
+    auto bj_lin_op = bj_factory->generate(std::move(mtx));
+    auto bj = static_cast<Bj *>(bj_lin_op.get());
+    d_bj_factory = BjFactory::create(gpu, 3);
+    auto d_bj_lin_op = d_bj_factory->generate(std::move(d_mtx));
+    auto d_bj = static_cast<Bj *>(d_bj_lin_op.get());
+
+    ASSERT_EQ(d_bj->get_num_blocks(), bj->get_num_blocks());
+    EXPECT_EQ(d_bj->get_max_block_size(), bj->get_max_block_size());
+}
+
+
+TEST_F(BlockJacobi, GpuFindNaturalBlocksInLargeMatrixEquivalentToRef)
+{
+    /* example matrix:
+        1   1
+        1   1
+        1       1
+        1       1
+        1       1
+        1       1
+     */
+    using data = gko::matrix_data<double, int>;
+    using nnz = data::nonzero_type;
+    auto m = data::diag(gko::dim{550, 550}, {{1.0, 1.0, 0.0, 0.0, 0.0, 0.0},
+                                             {1.0, 1.0, 0.0, 0.0, 0.0, 0.0},
+                                             {1.0, 0.0, 1.0, 0.0, 0.0, 0.0},
+                                             {1.0, 0.0, 1.0, 0.0, 0.0, 0.0},
+                                             {1.0, 0.0, 1.0, 0.0, 0.0, 0.0},
+                                             {1.0, 0.0, 1.0, 0.0, 0.0, 0.0}});
+    auto mtx = Mtx::create(ref);
+    mtx->read(m);
+    auto d_mtx = Mtx::create(gpu);
+    d_mtx->copy_from(mtx.get());
+
+    std::unique_ptr<BjFactory> bj_factory;
+    bj_factory = BjFactory::create(ref, 3);
+    auto bj_lin_op = bj_factory->generate(std::move(mtx));
+    auto bj = static_cast<Bj *>(bj_lin_op.get());
+    d_bj_factory = BjFactory::create(gpu, 3);
+    auto d_bj_lin_op = d_bj_factory->generate(std::move(d_mtx));
+    auto d_bj = static_cast<Bj *>(d_bj_lin_op.get());
+    ASSERT_EQ(d_bj->get_num_blocks(), bj->get_num_blocks());
+    EXPECT_EQ(d_bj->get_max_block_size(), bj->get_max_block_size());
+}
+
+
+TEST_F(BlockJacobi,
+       GpuExecutesSupervariableAgglomerationInLargeMatrixEquivalentToRef)
+{
+    /* example matrix:
+        1   1
+        1   1
+                1   1
+                1   1
+                        1
+     */
+    using data = gko::matrix_data<double, int>;
+    using nnz = data::nonzero_type;
+    auto m = data::diag(gko::dim{550, 550}, {{1.0, 1.0, 0.0, 0.0, 0.0},
+                                             {1.0, 1.0, 0.0, 0.0, 0.0},
+                                             {0.0, 0.0, 1.0, 1.0, 0.0},
+                                             {0.0, 0.0, 1.0, 1.0, 0.0},
+                                             {0.0, 0.0, 0.0, 0.0, 1.0}});
+    auto mtx = Mtx::create(ref);
+    mtx->read(m);
+    auto d_mtx = Mtx::create(gpu);
+    d_mtx->copy_from(mtx.get());
+
+    std::unique_ptr<BjFactory> bj_factory;
+    bj_factory = BjFactory::create(ref, 3);
```

Comment:
nit: not an error, but why not just `auto bj_factory = BjFactory::create(ref, 3)` instead of the two lines?

There's no need to ever specify the return type with Ginkgo (+ declaring variables as late as possible is considered good practice in C++).